pipelines/batch/src/main/java/org/openmrs/analytics/JdbcFetchHapi.java

```diff
@@ -103,29 +108,71 @@ public FetchRowsJdbcIo(JdbcIO.DataSourceConfiguration dataSourceConfig, String s
-      return queryParameters.apply(
-          "JdbcIO readAll",
-          JdbcIO.<QueryParameterDescriptor, HapiRowDescriptor>readAll()
-              .withDataSourceConfiguration(dataSourceConfig)
-              .withParameterSetter(
-                  new JdbcIO.PreparedStatementSetter<QueryParameterDescriptor>() {
-
-                    @Override
-                    public void setParameters(
-                        QueryParameterDescriptor element, PreparedStatement preparedStatement)
-                        throws Exception {
-                      preparedStatement.setString(1, element.resourceType());
-                      preparedStatement.setInt(2, element.numBatches());
-                      preparedStatement.setInt(3, element.batchId());
+      return queryParameters
+          .apply(
+              "JdbcIO readAll",
+              JdbcIO.<QueryParameterDescriptor, HapiRowDescriptor>readAll()
+                  .withDataSourceConfiguration(dataSourceConfig)
+                  .withParameterSetter(
+                      new JdbcIO.PreparedStatementSetter<QueryParameterDescriptor>() {
+
+                        @Override
+                        public void setParameters(
+                            QueryParameterDescriptor element, PreparedStatement preparedStatement)
+                            throws Exception {
+                          preparedStatement.setString(1, element.resourceType());
+                          preparedStatement.setInt(2, element.numBatches());
+                          preparedStatement.setInt(3, element.batchId());
+                        }
+                      })
+                  // We are disabling this parameter because by default, this parameter causes
+                  // JdbcIO to
+                  // add a reshuffle transform after reading from the database. This breaks fusion
+                  // between the read and write operations, thus resulting in high memory overhead.
+                  // Disabling the below parameter results in optimal performance.
+                  .withOutputParallelization(false)
+                  .withQuery(query)
+                  .withRowMapper(new ResultSetToRowDescriptor()))
+          .apply(
+              "Jdbc fetch tags",
+              ParDo.of(
+                  new DoFn<HapiRowDescriptor, HapiRowDescriptor>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext processContext) throws SQLException {
+                      HapiRowDescriptor hapiRowDescriptor = processContext.element();
+                      String resourceId = hapiRowDescriptor.resourceId();
+                      if (Strings.isNullOrEmpty(resourceId)) {
+                        throw new IllegalStateException("resourceId cant be null or empty");
+                      }
+                      List<Coding> tags = getResourceMetaTags(Integer.parseInt(resourceId));
```

Comment:
Can you please do a batch pipeline evaluation with and without the tags and record the difference somewhere (preferably on the bug)? You can use one of the usual test datasets we use, e.g., the 8K or 80K patients.

Reply:
sure, will do that.

Reply:
Added the observation in the issue (https://github.com/google/fhir-data-pipes/issues/563). This is for 8k patients.

Reply:
resolving this for now